cli/npm/resolution/graph.rs

```diff
@@ -487,200 +705,252 @@ impl<'a, TNpmRegistryApi: NpmRegistryApi>
+            match self.api.package_version_info(&pkg_nv).await? {
+              Some(version_info) => {
+                self.dep_entry_cache.store(pkg_nv.clone(), &version_info)?
+              }
+              None => {
+                bail!("Could not find version information for {}", pkg_nv)
+              }
+            }
+          };
+
+          (pkg_nv, deps)
         };
 
         // cache all the dependencies' registry infos in parallel if should
-        if !should_sync_download() {
-          let handles = deps
-            .iter()
-            .map(|dep| {
-              let name = dep.name.clone();
-              let api = self.api.clone();
-              tokio::task::spawn(async move {
-                // it's ok to call this without storing the result, because
-                // NpmRegistryApi will cache the package info in memory
-                api.package_info(&name).await
-              })
-            })
-            .collect::<Vec<_>>();
-          let results = futures::future::join_all(handles).await;
-          for result in results {
-            result??; // surface the first error
-          }
-        }
+        self
+          .api
+          .cache_in_parallel({
+            deps.iter().map(|dep| dep.name.clone()).collect()
+          })
+          .await?;
 
         // resolve the dependencies
         let mut found_peer = false;
+
         for dep in deps.iter() {
           let package_info = self.api.package_info(&dep.name).await?;
 
           match dep.kind {
             NpmDependencyEntryKind::Dep => {
-              let node = self.analyze_dependency(
-                dep,
-                &package_info,
-                &parent_id,
-                &visited_versions,
-              )?;
+              let maybe_child_id = self
+                .graph
+                .nodes
+                .get(&graph_path.node_id())
+                .unwrap()
+                .children
+                .get(&dep.bare_specifier)
+                .copied();
+              let child_id = if let Some(child_id) = maybe_child_id {
+                // we already resolved this dependency before, and we can skip over it
+                self.try_add_pending_unresolved_node(&graph_path, child_id);
+                child_id
+              } else {
+                self.analyze_dependency(dep, &package_info, &graph_path)?
+              };
+
               if !found_peer {
-                found_peer = !node.lock().no_peers;
+                found_peer = !self.graph.borrow_node_mut(child_id).no_peers;
               }
             }
             NpmDependencyEntryKind::Peer
             | NpmDependencyEntryKind::OptionalPeer => {
               found_peer = true;
-              let maybe_new_parent_id = self.resolve_peer_dep(
+              // we need to re-evaluate peer dependencies every time and can't
+              // skip over them because they might be evaluated differently based
+              // on the current treat
```

Comment:
current treat?

Reply:
"current path". Maybe I was thinking about food at the time.